factor/totp/classes/factor.php

```diff
@@ -263,21 +259,22 @@ public function has_revoke() {
      *
      * {@inheritDoc}
      */
-    public function has_lastverified() {
+    public function has_setup() {
         return true;
     }
 
+    public function get_device_name($factorid) {
```

Comment:
Does this make sense any more? It is now shared and called label, so I think move this to the base class and call it get_label()

...but I'm not convinced we need it at all. We can probably just remove it completely

Reply:
Its current use case is just for action.php, but I think it is reasonable to broaden the scope of it a bit and move it to the base class. It is currently only used in one place, but it will most likely become useful in future factors, most likely email.